File: python/atelier_core/policy.py
"""Authorisation policy using structural pattern matching on the actor's role."""

from __future__ import annotations

from dataclasses import dataclass

from .models.repair_order import RepairOrder
from .support.status import RepairStatus


@dataclass(frozen=True)
class Actor:
    id: int
    role: str


class RepairOrderPolicy:
    def can_view(self, actor: Actor, order: RepairOrder) -> bool:
        return actor.id == order.customer_id if actor.role == "customer" else True

    def can_transition(self, actor: Actor, order: RepairOrder) -> bool:
        if order.status.is_terminal():
            return False
        return actor.role != "customer"

    def can_collect(self, actor: Actor, order: RepairOrder) -> bool:
        """`match` with class patterns and guards; no default arm returns False."""
        if order.status is not RepairStatus.COMPLETED:
            return False
        match actor:
            case Actor(role="customer", id=customer_id) if customer_id == order.customer_id:
                return True
            case Actor(role="manager"):
                return True
            case _:
                return False

File: go/core/policy/policy.go
// Package policy answers "may this actor do this", the way a framework gate
// would: a method per ability, dispatched by a string name.
package policy

import (
	"atelier.example/lane/core/models"
	"atelier.example/lane/core/support"
)

// Actor is whoever is asking.
type Actor struct {
	ID   int
	Role string
}

// NewActor builds an actor.
func NewActor(id int, role string) Actor {
	return Actor{ID: id, Role: role}
}

// RepairOrderPolicy carries the abilities for one model.
type RepairOrderPolicy struct{}

// NewRepairOrderPolicy builds the policy.
func NewRepairOrderPolicy() *RepairOrderPolicy {
	return &RepairOrderPolicy{}
}

// CanView is open to the owner and to staff.
func (p *RepairOrderPolicy) CanView(actor Actor, order *models.RepairOrder) bool {
	return actor.Role != "guest" || actor.ID == order.CustomerID
}

// CanCollect requires a completed order and the owning customer.
func (p *RepairOrderPolicy) CanCollect(actor Actor, order *models.RepairOrder) bool {
	if order.Status != support.StatusCompleted {
		return false
	}
	return actor.ID == order.CustomerID
}

// CanEdit is staff only.
func (p *RepairOrderPolicy) CanEdit(actor Actor, order *models.RepairOrder) bool {
	return actor.Role == "technician" || actor.Role == "manager"
}

// Allows dispatches an ability by STRING NAME, which is the edge a static
// tool has to guess: the method is chosen from a value, not from a selector.
func (p *RepairOrderPolicy) Allows(ability string, actor Actor, order *models.RepairOrder) bool {
	switch ability {
	case "view":
		return p.CanView(actor, order)
	case "collect":
		return p.CanCollect(actor, order)
	case "edit":
		return p.CanEdit(actor, order)
	default:
		return false
	}
}
